testament/backend.nim

```diff
@@ -1,70 +1,121 @@
+      close resFile
+    currentCategory = category
+    results = newJArray()
+
+  results.add %*{"name": name, "category": category, "target": target,
+    "action": action, "result": result, "expected": expected, "given": given,
+    "machine": thisMachine.string, "commit": thisCommit.string, "branch": thisBranch}
+
+proc open*() =
+  thisMachine = getMachine()
+  thisCommit = getCommit()
+
+proc close*() =
+  if currentCategory.len > 0:
+    var resFile = open("testresults" / currentCategory.addFileExt"json", fmWrite)
+    resFile.write pretty(results)
+    close resFile
+
+proc cacheResults*() =
+  ## Traverses the testresults directory and extracts any failed json entries
+  ## and inserts them into a new file within the cacheresults directory.
+  createDir("testresults/cacheresults")
+  # We will ignore any entries that have this as their result
+  const passResults = ["reJoined", "reSuccess", "reDisabled", ""] # "" is defaulted to if result field not found
+  let searchPattern = "testresults" & DirSep & "*.json"
+  # Prepare json array which will be written to our new cache file
+  var fresults = newJArray()
+  # Walk files in testresults that have a json ext
+  for f in walkFiles(searchPattern):
+    var jdata: JsonNode
+    try:
+      # Parse the json file into the json node
+        jdata = f.parseFile()
+    except IOError, JsonParsingError:
+      echo getCurrentExceptionMsg()
+      jdata = newJArray()
+    # The testresults files should be made up of arrays of json objects representing
+    # an entry each. Traverse the array and check the result field of each object.
+    # If the result is not one of the "passResults", then we add it to our cache
+    # node 'fresults'
+    for node in jdata:
+      let noderesult = node{"result"}.getStr()
+      if not passResults.contains(noderesult):
+        fresults.add(node)
+  
+  # if len(fresults) > 0: # We want to overwrite the cache if we have a succesful
+  # run, therefore we should write to the file even if the cache node is empty
+  var results = open("testresults" / "cacheresults" / "result".addFileExt"json", fmWrite)
+  results.write(fresults.pretty())
+  close(results)
+
+proc getRetries*(): (seq[string], seq[(string, string)]) = # Returns all the categories of concern and then the "name" entry list
+  ## Result = (Categories that had fails, Name and target combination of failed test)
+  # Gets the files that failed in the previous run from the cache
+  var cats: seq[string]
+  var file_targs: seq[(string, string)]
+  const cacheDir = "testresults" / "cacheresults"
+  createDir(cacheDir)
+  const cacheFile = cacheDir / "result".addFileExt"json"
+  let cacheArray = cacheFile.parseFile()
+  # Traverse the array of entries, add the category to cats if it has not been seen,
+  # either way add the "name" field to file_targs
+  for entry in cacheArray:
+    let cat = entry["category"].getStr()
+    let file_targ = entry["name"].getStr()
+    let name_target = file_targ.split()

```

Comment:
Are you trying to split `tests/something/foo.nim C` into path and target? This kind of splitting will not work with the `dir with spaces` tests. Probably an `rsplit` with `maxsplit` of 2 would work.